apps/labs/posts/narwhals-pycapsule.md

```diff
@@ -0,0 +1,199 @@
+---
+title: 'Universal dataframe support with the Arrow PyCapsule Interface + Narwhals'
+published: January 6, 2025
+authors: [marco-gorelli]
+description: 'Support everything, depend on (almost) nothing'
+category: [PyData ecosystem]
+featuredImage:
+  src: /posts/narwhals-pycapsule/featured.jpg
+  alt: 'Cat looking out window, with dataframe logos outside, original image by  Lucy Jackline https://unsplash.com/photos/a-cat-sitting-on-a-window-sill-looking-out-a-window-O896LIqr2vc'
+hero:
+  imageSrc: /posts/narwhals-pycapsule/hero.jpg
+  imageAlt: 'Cat looking out window, with dataframe logos outside, original image by  Lucy Jackline https://unsplash.com/photos/a-cat-sitting-on-a-window-sill-looking-out-a-window-O896LIqr2vc'
+---
+
+# Universal dataframe support with the Arrow PyCapsule Interface + Narwhals
+
+If you were writing a data science tool in 2015, you'd probably have
+added pandas support and called it a day. However, it's not 2015, we've fast-forwarded to 2025,
+and so if you do that today you'll be inundated with endless "can you support
+Polars / DuckDB / PyArrow / DataFusion / ..." requests. Yet you have no interest in understanding
+the subtle differences between them, prefering to focus on the problems which your library set
+out to solve. What can you do?
+
+Today, you'll learn about how to create tools which support all kinds of dataframes:
+
+- The [Arrow PyCapsule Interface](https://arrow.apache.org/docs/format/CDataInterface/PyCapsuleInterface.html),
+  if you need access to dataframe data from a low-level language like C or Rust.
+- [Narwhals](https://github.com/narwhals-dev/narwhals), if you want to keep your logic in Python.
+- Narwhals _and_ the PyCapsule Interface together, if you want it all!
+
+## `agnostic_sum_i64_column` - "slow down, Professor!"
+
+We'll learn about how you make a dataframe-agnostic function `agnostic_sum_i64_column`.
+Here are the requirements:
+
+- Given a dataframe `df` and a column name `column_name`, compute
+  the sum of all non-null values in that column.
+- If that column is not present, or if it is not of type `Int64`,
+  raise an error.
+- `pandas.DataFrame`, `polars.DataFrame`,
+  `duckdb.PyRelation`, and `pyarrow.Table`, should all be supported without
+  any of them being required.
+
+Summing non-null values isn't rocket science, but how can we do this in a way that all
+input libraries are supported without any of them being required?
+
+## Low-level solution: PyCapsule Interface in Rust via PyO3
+
+An example of a Rust solution via Pyo3 can be found at [pycapsule-demo/src/lib.rs](https://github.com/MarcoGorelli/pycapsule-demo/blob/6aad64be26ebbfc8526f26695544bfc6436e3266/src/lib.rs#L9-L56).
+
+The technical details are beyond the scope of this post, but the summary is:
+
+- We accept any object which implements the
```

Comment:
Not sure how technical you want to be and I realize this post is aimed at dataframe libraries, but there are other methods that can be implemented that still fall under the PyCapsule Interface

 * __arrow_c_array__
 * __arrow_c_schema__
 * __arrow_c_stream__
 * __arrow_c_device_array__
 * __arrow_device_array_stream__

Stream may be the most fitting for exchanging dataframes, but the other ones have use outside of 2-d structures. May want to be careful in saying that the presence of this method exclusively means you are implementing the PyCapsule interface